File: src/gui/settlement_loader.py
"""
Settlement boundary loader — fetches from Overpass API with tile caching.
"""
import json
import math
import os
import threading
import time
import urllib.request
import urllib.parse

from PyQt5.QtCore import QObject, pyqtSignal, QThread


_TILE_GRID = 0.1  # ~10 km tile grid
_CACHE_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', 'cache', 'settlements_v5')
_DP_TOLERANCE = 0.0005  # ~55 m - Douglas-Peucker simplification

_OVERPASS_ENDPOINTS = [
    'https://overpass-api.de/api/interpreter',
    'https://overpass.kumi.systems/api/interpreter',
]


def _close(a, b, eps=1e-6):
    return abs(a[0] - b[0]) < eps and abs(a[1] - b[1]) < eps


class _FetchWorker(QThread):
    """Batch-fetches settlements for all queued tiles in one Overpass request."""
    tile_loaded = pyqtSignal(str, list)

    def __init__(self, loader):
        super().__init__()
        self._loader = loader

    def run(self):
        while True:
            with self._loader._lock:
                if not self._loader._queue:
                    return
                tiles = list(self._loader._queue)
                self._loader._queue.clear()

            s = min(t[0] for t in tiles)
            w = min(t[1] for t in tiles)
            n = max(t[2] for t in tiles)
            e = max(t[3] for t in tiles)

            query = (
                f'[out:json][timeout:30];('
                f'way["place"~"^(city|town|village|hamlet)$"]({s},{w},{n},{e});'
                f'relation["place"~"^(city|town|village|hamlet)$"]({s},{w},{n},{e});'
                f'node["place"~"^(city|town|village|hamlet)$"]({s},{w},{n},{e});'
                f');out geom;'
            )

            last_err = None
            for attempt in range(len(_OVERPASS_ENDPOINTS)):
                try:
                    ep = _OVERPASS_ENDPOINTS[attempt]
                    post_data = urllib.parse.urlencode({'data': query}).encode()
                    req = urllib.request.Request(ep, data=post_data)
                    with urllib.request.urlopen(req, timeout=30) as resp:
                        raw = json.loads(resp.read())

                    elements = raw.get('elements', [])
                    all_features = SettlementLoader._process_elements(elements)
                    tile_map = self._distribute(all_features, tiles)

                    with self._loader._lock:
                        for key, features in tile_map.items():
                            self._loader._cache[key] = features
                            self._loader._emitted_keys.add(key)

                    for key, features in tile_map.items():
                        SettlementLoader._save_tile(key, features)
                        self.tile_loaded.emit(key, features)

                    last_err = None
                    break
                except Exception as exc:
                    last_err = exc
                    if attempt < len(_OVERPASS_ENDPOINTS) - 1:
                        time.sleep(2)

            if last_err:
                print(f"[FIX] Settlement batch failed: {last_err}")
                with self._loader._lock:
                    for t in tiles:
                        self._loader._loaded_keys.discard(SettlementLoader._key(t))

    @staticmethod
    def _distribute(features, tiles):
        """Assign features to overlapping tiles by bbox."""
        tile_map = {}
        for t in tiles:
            tile_map[SettlementLoader._key(t)] = []
        for f in features:
            if f['F'] == 'p':
                lats = [c[0] for c in f['c']]
                lons = [c[1] for c in f['c']]
                fb = (min(lats), min(lons), max(lats), max(lons))
            else:
                fb = (f['lat'], f['lon'], f['lat'], f['lon'])
            for t in tiles:
                if not (fb[2] < t[0] or fb[0] > t[2] or fb[3] < t[1] or fb[1] > t[3]):
                    tile_map[SettlementLoader._key(t)].append(f)
        return tile_map


class SettlementLoader(QObject):
    """Batch-fetches settlement boundaries with tile caching."""
    tile_loaded = pyqtSignal(str, list)

    def __init__(self):
        super().__init__()
        self._queue = []
        self._loaded_keys = set()
        self._emitted_keys = set()
        self._cache = {}
        self._lock = threading.Lock()
        os.makedirs(_CACHE_DIR, exist_ok=True)
        self._load_disk_cache()
        self._worker = _FetchWorker(self)
        self._worker.tile_loaded.connect(self.tile_loaded)
        self._worker.finished.connect(self._on_worker_done)

    def request(self, south, west, north, east):
        tiles = self._grid_tiles(south, west, north, east)
        added = False
        with self._lock:
            for t in tiles:
                key = self._key(t)
                if key not in self._loaded_keys:
                    self._loaded_keys.add(key)
                    self._queue.append(t)
                    added = True
        if added:
            self._kick_worker()

    def _kick_worker(self):
        if not self._worker.isRunning():
            self._worker.start()

    def _on_worker_done(self):
        with self._lock:
            has_work = bool(self._queue)
        if has_work:
            self._worker.start()

    @staticmethod
    def _process_elements(elements):
        """Extract polygons from ways/relations, fall back to nodes for the rest."""
        polys = []
        nodes = []
        poly_bboxes = []

        for el in elements:
            place = el.get('tags', {}).get('place', '')
            if not place:
                continue

            if el['type'] == 'way' and 'geometry' in el:
                coords = [[p['lat'], p['lon']] for p in el['geometry']]
                if len(coords) < 3:
                    continue
                lats = [c[0] for c in coords]
                lons = [c[1] for c in coords]
                span = max(max(lats) - min(lats), max(lons) - min(lons))
                tol = _DP_TOLERANCE
                if span > 0.5:
                    tol = 0.005
                elif span > 0.2:
                    tol = 0.002
                elif span > 0.1:
                    tol = 0.001
                simplified = SettlementLoader._simplify_dp(coords, tol)
                polys.append({'F': 'p', 'c': simplified, 't': place})
                poly_bboxes.append((min(lats), min(lons), max(lats), max(lons)))

            elif el['type'] == 'relation' and 'members' in el:
                ring = SettlementLoader._merge_relation(el['members'])
                if ring and len(ring) >= 3:
                    lats = [c[0] for c in ring]
                    lons = [c[1] for c in ring]
                    span = max(max(lats) - min(lats), max(lons) - min(lons))
                    tol = _DP_TOLERANCE
                    if span > 0.5:
                        tol = 0.005
                    elif span > 0.2:
                        tol = 0.002
                    elif span > 0.1:
                        tol = 0.001
                    simplified = SettlementLoader._simplify_dp(ring, tol)
                    polys.append({'F': 'p', 'c': simplified, 't': place})
                    poly_bboxes.append((min(lats), min(lons), max(lats), max(lons)))

            elif el['type'] == 'node':
                nodes.append({'F': 'n', 'lat': el['lat'], 'lon': el['lon'], 't': place})

        # Filter out nodes that already have a polygon
        result = list(polys)
        for nd in nodes:
            covered = False
            for (mn_la, mn_lo, mx_la, mx_lo) in poly_bboxes:
                if mn_la <= nd['lat'] <= mx_la and mn_lo <= nd['lon'] <= mx_lo:
                    covered = True
                    break
            if not covered:
                result.append(nd)
        return result

    @staticmethod
    def _merge_relation(members):
        """Merge outer way geometries of a relation into a single ring."""
        segments = []
        for m in members:
            if m.get('type') != 'way':
                continue
            role = m.get('role', 'outer')
            if role not in ('outer', ''):
                continue
            geom = m.get('geometry')
            if not geom:
                continue
            seg = [[p['lat'], p['lon']] for p in geom]
            if seg:
                segments.append(seg)
        if not segments:
            return []
        ring = list(segments[0])
        remaining = segments[1:]
        max_iter = len(remaining) * 2
        i = 0
        while remaining and i < max_iter:
            i += 1
            matched = False
            for idx, seg in enumerate(remaining):
                if _close(ring[-1], seg[0]):
                    ring.extend(seg[1:])
                    remaining.pop(idx)
                    matched = True
                    break
                if _close(ring[-1], seg[-1]):
                    ring.extend(reversed(seg[:-1]))
                    remaining.pop(idx)
                    matched = True
                    break
            if not matched:
                break
        return ring

    @staticmethod
    def _simplify_dp(coords, tolerance):
        """Douglas-Peucker polyline simplification."""
        if len(coords) <= 4:
            return [[round(c[0], 5), round(c[1], 5)] for c in coords]

        def _perp_dist(p, a, b):
            dx, dy = b[0] - a[0], b[1] - a[1]
            if dx == 0 and dy == 0:
                return math.hypot(p[0] - a[0], p[1] - a[1])
            t = ((p[0] - a[0]) * dx + (p[1] - a[1]) * dy) / (dx * dx + dy * dy)
            t = max(0, min(1, t))
            proj = [a[0] + t * dx, a[1] + t * dy]
            return math.hypot(p[0] - proj[0], p[1] - proj[1])

        def _dp(pts, tol):
            if len(pts) <= 2:
                return pts
            max_d, max_i = 0, 0
            for i in range(1, len(pts) - 1):
                d = _perp_dist(pts[i], pts[0], pts[-1])
                if d > max_d:
                    max_d, max_i = d, i
            if max_d > tol:
                left = _dp(pts[:max_i + 1], tol)
                right = _dp(pts[max_i:], tol)
                return left[:-1] + right
            return [pts[0], pts[-1]]

        result = _dp(coords, tolerance)
        return [[round(c[0], 5), round(c[1], 5)] for c in result]

    @staticmethod
    def _key(tile):
        return f"{tile[0]},{tile[1]}"

    @staticmethod
    def _grid_tiles(south, west, north, east):
        g = _TILE_GRID
        tiles = []
        s = math.floor(south / g) * g
        while s < north:
            w_cur = math.floor(west / g) * g
            while w_cur < east:
                tiles.append((
                    round(s, 4), round(w_cur, 4),
                    round(s + g, 4), round(w_cur + g, 4),
                ))
                w_cur += g
            s += g
        return tiles

    def _load_disk_cache(self):
        try:
            for fname in os.listdir(_CACHE_DIR):
                if not fname.endswith('.json'):
                    continue
                key = fname[:-5]
                path = os.path.join(_CACHE_DIR, fname)
                with open(path, 'r', encoding='utf-8') as f:
                    self._cache[key] = json.load(f)
                self._loaded_keys.add(key)
        except Exception as e:
            print(f"Settlement cache load error: {e}")

    @staticmethod
    def _save_tile(key, features):
        try:
            path = os.path.join(_CACHE_DIR, key + '.json')
            with open(path, 'w', encoding='utf-8') as f:
                json.dump(features, f, ensure_ascii=False, separators=(',', ':'))
        except Exception as e:
            print(f"Settlement cache save error: {e}")

    def preload_cached_tiles(self, south, west, north, east):
        tiles = self._grid_tiles(south, west, north, east)
        for t in tiles:
            key = self._key(t)
            if key in self._cache and self._cache[key] and key not in self._emitted_keys:
                self._emitted_keys.add(key)
                self.tile_loaded.emit(key, self._cache[key])

File: cpp/src/gui/SettlementLoader.cpp
#include "SettlementLoader.h"
#include <QDir>
#include <QFile>
#include <QStandardPaths>
#include <QCoreApplication>
#include <QNetworkAccessManager>
#include <QNetworkRequest>
#include <QNetworkReply>
#include <QEventLoop>
#include <QUrlQuery>
#include <cmath>
#include <algorithm>
#include <spdlog/spdlog.h>

namespace vtol {

static const QStringList OVERPASS_ENDPOINTS = {
    QStringLiteral("https://overpass-api.de/api/interpreter"),
    QStringLiteral("https://overpass.kumi.systems/api/interpreter"),
};

// ═════════════════════════════════════════════════════════════════════════
//  SettlementFetchWorker
// ═════════════════════════════════════════════════════════════════════════

SettlementFetchWorker::SettlementFetchWorker(SettlementLoader* loader)
    : QThread(loader), m_loader(loader) {}

void SettlementFetchWorker::run()
{
    while (true) {
        std::vector<TileBounds> tiles;
        {
            QMutexLocker lk(&m_loader->m_mutex);
            if (m_loader->m_queue.empty()) return;
            // Limit batch size to avoid Overpass timeout on large bbox
            constexpr size_t MAX_BATCH = 6;
            if (m_loader->m_queue.size() <= MAX_BATCH) {
                tiles = std::move(m_loader->m_queue);
                m_loader->m_queue.clear();
            } else {
                tiles.assign(m_loader->m_queue.begin(),
                             m_loader->m_queue.begin() + MAX_BATCH);
                m_loader->m_queue.erase(m_loader->m_queue.begin(),
                                         m_loader->m_queue.begin() + MAX_BATCH);
            }
        }

        // Compute merged bbox
        double s = 90, w = 180, n = -90, e = -180;
        for (auto& [ts, tw, tn, te] : tiles) {
            s = std::min(s, ts); w = std::min(w, tw);
            n = std::max(n, tn); e = std::max(e, te);
        }

        auto query = QString(
            "[out:json][timeout:30];("
            "way[\"place\"~\"^(city|town|village|hamlet)$\"](%1,%2,%3,%4);"
            "relation[\"place\"~\"^(city|town|village|hamlet)$\"](%1,%2,%3,%4);"
            "node[\"place\"~\"^(city|town|village|hamlet)$\"](%1,%2,%3,%4);"
            ");out geom;")
            .arg(s, 0, 'f', 6).arg(w, 0, 'f', 6)
            .arg(n, 0, 'f', 6).arg(e, 0, 'f', 6);

        SPDLOG_INFO("[SettlementLoader] Fetching {} tiles, bbox [{:.4f},{:.4f},{:.4f},{:.4f}]",
                    tiles.size(), s, w, n, e);

        // Try each endpoint
        bool success = false;
        for (int attempt = 0; attempt < OVERPASS_ENDPOINTS.size(); ++attempt) {
            try {
                QNetworkAccessManager nam;
                QNetworkRequest req{QUrl{OVERPASS_ENDPOINTS[attempt]}};
                req.setHeader(QNetworkRequest::ContentTypeHeader,
                              "application/x-www-form-urlencoded");

                QUrlQuery postData;
                postData.addQueryItem("data", query);

                QEventLoop loop;
                auto* reply = nam.post(req, postData.toString(QUrl::FullyEncoded).toUtf8());
                QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
                loop.exec();

                if (reply->error() != QNetworkReply::NoError) {
                    SPDLOG_WARN("[SettlementLoader] HTTP error: {}",
                                reply->errorString().toStdString());
                    reply->deleteLater();
                    if (attempt < OVERPASS_ENDPOINTS.size() - 1)
                        QThread::sleep(2);
                    continue;
                }

                auto raw = nlohmann::json::parse(reply->readAll().toStdString());
                reply->deleteLater();

                auto elements = raw.value("elements", nlohmann::json::array());
                auto allFeatures = SettlementLoader::processElements(elements);
                auto tileMap = distribute(allFeatures, tiles);

                {
                    QMutexLocker lk(&m_loader->m_mutex);
                    for (auto it = tileMap.begin(); it != tileMap.end(); ++it) {
                        m_loader->m_cache[it.key()] = it.value();
                        m_loader->m_emittedKeys.insert(it.key());
                    }
                }

                for (auto it = tileMap.begin(); it != tileMap.end(); ++it) {
                    SettlementLoader::saveTile(it.key(), it.value());
                    emit tileLoaded(it.key(), it.value());
                }

                SPDLOG_DEBUG("[SettlementLoader] Fetched {} features → {} tiles",
                             allFeatures.size(), tileMap.size());
                success = true;
                break;

            } catch (const std::exception& ex) {
                SPDLOG_WARN("[SettlementLoader] Attempt {} failed: {}",
                            attempt + 1, ex.what());
                if (attempt < OVERPASS_ENDPOINTS.size() - 1)
                    QThread::sleep(2);
            }
        }

        if (!success) {
            SPDLOG_ERROR("[SettlementLoader] All endpoints failed");
            QMutexLocker lk(&m_loader->m_mutex);
            for (auto& t : tiles)
                m_loader->m_loadedKeys.remove(SettlementLoader::tileKey(t));
        }
    }
}

QMap<QString, FeatureList>
SettlementFetchWorker::distribute(const FeatureList& features,
                                   const std::vector<TileBounds>& tiles)
{
    QMap<QString, FeatureList> result;
    for (auto& t : tiles)
        result[SettlementLoader::tileKey(t)] = {};

    for (auto& f : features) {
        double fs, fw, fn, fe;
        if (f.form == 'p') {
            fs = fn = f.coords[0][0];
            fw = fe = f.coords[0][1];
            for (auto& c : f.coords) {
                fs = std::min(fs, c[0]); fn = std::max(fn, c[0]);
                fw = std::min(fw, c[1]); fe = std::max(fe, c[1]);
            }
        } else {
            fs = fn = f.lat;
            fw = fe = f.lon;
        }
        for (auto& t : tiles) {
            auto [ts, tw, tn, te] = t;
            if (!(fn < ts || fs > tn || fe < tw || fw > te))
                result[SettlementLoader::tileKey(t)].push_back(f);
        }
    }
    return result;
}

// ═════════════════════════════════════════════════════════════════════════
//  SettlementLoader
// ═════════════════════════════════════════════════════════════════════════

SettlementLoader::SettlementLoader(QObject* parent)
    : QObject(parent)
{
    m_cacheDir = QCoreApplication::applicationDirPath() + "/cache/settlements_v5";
    QDir().mkpath(m_cacheDir);

    m_worker = new SettlementFetchWorker(this);
    connect(m_worker, &SettlementFetchWorker::tileLoaded,
            this, &SettlementLoader::tileLoaded);
    connect(m_worker, &QThread::finished, this, &SettlementLoader::onWorkerDone);

    loadDiskCache();
}

void SettlementLoader::request(double south, double west, double north, double east)
{
    auto tiles = gridTiles(south, west, north, east);
    bool added = false;

    {
        QMutexLocker lk(&m_mutex);
        for (auto& t : tiles) {
            auto key = tileKey(t);
            if (!m_loadedKeys.contains(key)) {
                m_loadedKeys.insert(key);
                m_queue.push_back(t);
                added = true;
            }
        }
    }

    if (added) kickWorker();

    // Emit cached tiles that haven't been emitted yet
    preloadCachedTiles(south, west, north, east);
}

void SettlementLoader::preloadCachedTiles(double south, double west,
                                           double north, double east)
{
    auto tiles = gridTiles(south, west, north, east);
    QMutexLocker lk(&m_mutex);
    for (auto& t : tiles) {
        auto key = tileKey(t);
        if (m_cache.contains(key) && !m_cache[key].empty()
            && !m_emittedKeys.contains(key))
        {
            m_emittedKeys.insert(key);
            auto features = m_cache[key];  // copy under lock
            lk.unlock();
            emit tileLoaded(key, features);
            lk.relock();
        }
    }
}

void SettlementLoader::kickWorker()
{
    if (!m_worker->isRunning())
        m_worker->start();
}

void SettlementLoader::onWorkerDone()
{
    QMutexLocker lk(&m_mutex);
    if (!m_queue.empty()) {
        lk.unlock();
        m_worker->start();
    }
}

void SettlementLoader::clearCache()
{
    QMutexLocker lk(&m_mutex);
    m_cache.clear();
    m_loadedKeys.clear();
    m_emittedKeys.clear();
    m_queue.clear();
    lk.unlock();

    // Remove all files from cache dir
    QDir dir(m_cacheDir);
    for (auto& f : dir.entryList({"*.json"}, QDir::Files))
        dir.remove(f);
    SPDLOG_INFO("[SettlementLoader] Cache cleared: {}", m_cacheDir.toStdString());
}

// ── Disk cache ──────────────────────────────────────────────────────────

void SettlementLoader::loadDiskCache()
{
    QDir dir(m_cacheDir);
    auto files = dir.entryList({"*.json"}, QDir::Files);
    for (auto& fname : files) {
        auto key = fname.left(fname.size() - 5); // strip .json
        try {
            QFile f(dir.filePath(fname));
            if (!f.open(QIODevice::ReadOnly)) continue;
            auto data = nlohmann::json::parse(f.readAll().toStdString());

            FeatureList features;
            for (auto& el : data) {
                SettlementFeature sf;
                sf.form = el.value("F", "p") == "p" ? 'p' : 'n';
                sf.type = el.value("t", "");
                if (sf.form == 'p') {
                    for (auto& c : el["c"])
                        sf.coords.push_back({c[0].get<double>(), c[1].get<double>()});
                } else {
                    sf.lat = el.value("lat", 0.0);
                    sf.lon = el.value("lon", 0.0);
                }
                features.push_back(std::move(sf));
            }
            m_cache[key] = std::move(features);
            m_loadedKeys.insert(key);
        } catch (const std::exception& ex) {
            SPDLOG_WARN("[SettlementLoader] Cache load error for {}: {}",
                        fname.toStdString(), ex.what());
        }
    }
    SPDLOG_INFO("[SettlementLoader] Loaded {} cached tiles", m_cache.size());
}

void SettlementLoader::saveTile(const QString& key, const FeatureList& features)
{
    try {
        // Use application dir path — but we're in a worker thread, so compose path
        auto dir = QCoreApplication::applicationDirPath() + "/cache/settlements_v5";
        QDir().mkpath(dir);
        auto path = dir + "/" + key + ".json";

        nlohmann::json arr = nlohmann::json::array();
        for (auto& f : features) {
            nlohmann::json obj;
            obj["F"] = (f.form == 'p') ? "p" : "n";
            obj["t"] = f.type;
            if (f.form == 'p') {
                nlohmann::json coords = nlohmann::json::array();
                for (auto& c : f.coords)
                    coords.push_back({std::round(c[0] * 1e5) / 1e5,
                                      std::round(c[1] * 1e5) / 1e5});
                obj["c"] = coords;
            } else {
                obj["lat"] = f.lat;
                obj["lon"] = f.lon;
            }
            arr.push_back(obj);
        }

        QFile file(path);
        if (file.open(QIODevice::WriteOnly)) {
            auto str = arr.dump(-1, ' ', false, nlohmann::json::error_handler_t::ignore);
            file.write(str.data(), static_cast<qint64>(str.size()));
        }
    } catch (const std::exception& ex) {
        SPDLOG_WARN("[SettlementLoader] Cache save error: {}", ex.what());
    }
}

// ── Processing ──────────────────────────────────────────────────────────

FeatureList SettlementLoader::processElements(const nlohmann::json& elements)
{
    FeatureList polys;
    std::vector<SettlementFeature> nodes;
    std::vector<std::array<double, 4>> polyBboxes; // s,w,n,e

    for (auto& el : elements) {
        auto place = el.value("/tags/place"_json_pointer, std::string());
        if (place.empty()) continue;

        auto type = el.value("type", "");

        if (type == "way" && el.contains("geometry")) {
            std::vector<std::array<double, 2>> coords;
            for (auto& p : el["geometry"])
                coords.push_back({p["lat"].get<double>(), p["lon"].get<double>()});
            if (coords.size() < 3) continue;

            double minLat = 90, maxLat = -90, minLon = 180, maxLon = -180;
            for (auto& c : coords) {
                minLat = std::min(minLat, c[0]); maxLat = std::max(maxLat, c[0]);
                minLon = std::min(minLon, c[1]); maxLon = std::max(maxLon, c[1]);
            }
            double span = std::max(maxLat - minLat, maxLon - minLon);
            double tol = DP_TOLERANCE;
            if (span > 0.5) tol = 0.005;
            else if (span > 0.2) tol = 0.002;
            else if (span > 0.1) tol = 0.001;

            SettlementFeature sf;
            sf.form = 'p';
            sf.coords = simplifyDP(coords, tol);
            sf.type = place;
            polys.push_back(std::move(sf));
            polyBboxes.push_back({minLat, minLon, maxLat, maxLon});

        } else if (type == "relation" && el.contains("members")) {
            auto ring = mergeRelation(el["members"]);
            if (ring.size() < 3) continue;

            double minLat = 90, maxLat = -90, minLon = 180, maxLon = -180;
            for (auto& c : ring) {
                minLat = std::min(minLat, c[0]); maxLat = std::max(maxLat, c[0]);
                minLon = std::min(minLon, c[1]); maxLon = std::max(maxLon, c[1]);
            }
            double span = std::max(maxLat - minLat, maxLon - minLon);
            double tol = DP_TOLERANCE;
            if (span > 0.5) tol = 0.005;
            else if (span > 0.2) tol = 0.002;
            else if (span > 0.1) tol = 0.001;

            SettlementFeature sf;
            sf.form = 'p';
            sf.coords = simplifyDP(ring, tol);
            sf.type = place;
            polys.push_back(std::move(sf));
            polyBboxes.push_back({minLat, minLon, maxLat, maxLon});

        } else if (type == "node") {
            SettlementFeature sf;
            sf.form = 'n';
            sf.lat = el["lat"].get<double>();
            sf.lon = el["lon"].get<double>();
            sf.type = place;
            nodes.push_back(std::move(sf));
        }
    }

    // Filter nodes covered by existing polygons
    FeatureList result = std::move(polys);
    for (auto& nd : nodes) {
        bool covered = false;
        for (auto& [s, w, n, e] : polyBboxes) {
            if (nd.lat >= s && nd.lat <= n && nd.lon >= w && nd.lon <= e) {
                covered = true;
                break;
            }
        }
        if (!covered)
            result.push_back(std::move(nd));
    }
    return result;
}

std::vector<std::array<double, 2>>
SettlementLoader::mergeRelation(const nlohmann::json& members)
{
    std::vector<std::vector<std::array<double, 2>>> segments;

    for (auto& m : members) {
        if (m.value("type", "") != "way") continue;
        auto role = m.value("role", "outer");
        if (role != "outer" && !role.empty()) continue;
        if (!m.contains("geometry")) continue;

        std::vector<std::array<double, 2>> seg;
        for (auto& p : m["geometry"])
            seg.push_back({p["lat"].get<double>(), p["lon"].get<double>()});
        if (!seg.empty())
            segments.push_back(std::move(seg));
    }

    if (segments.empty()) return {};

    auto ring = segments[0];
    std::vector<std::vector<std::array<double, 2>>> remaining(
        segments.begin() + 1, segments.end());

    auto close = [](const std::array<double, 2>& a, const std::array<double, 2>& b) {
        return std::abs(a[0] - b[0]) < 1e-6 && std::abs(a[1] - b[1]) < 1e-6;
    };

    int maxIter = static_cast<int>(remaining.size()) * 2;
    for (int i = 0; i < maxIter && !remaining.empty(); ++i) {
        bool matched = false;
        for (size_t idx = 0; idx < remaining.size(); ++idx) {
            if (close(ring.back(), remaining[idx].front())) {
                ring.insert(ring.end(), remaining[idx].begin() + 1, remaining[idx].end());
                remaining.erase(remaining.begin() + static_cast<ptrdiff_t>(idx));
                matched = true;
                break;
            }
            if (close(ring.back(), remaining[idx].back())) {
                auto& seg = remaining[idx];
                for (int j = static_cast<int>(seg.size()) - 2; j >= 0; --j)
                    ring.push_back(seg[static_cast<size_t>(j)]);
                remaining.erase(remaining.begin() + static_cast<ptrdiff_t>(idx));
                matched = true;
                break;
            }
        }
        if (!matched) break;
    }
    return ring;
}

std::vector<std::array<double, 2>>
SettlementLoader::simplifyDP(const std::vector<std::array<double, 2>>& coords,
                              double tolerance)
{
    if (coords.size() <= 4) {
        std::vector<std::array<double, 2>> out;
        out.reserve(coords.size());
        for (auto& c : coords)
            out.push_back({std::round(c[0] * 1e5) / 1e5,
                           std::round(c[1] * 1e5) / 1e5});
        return out;
    }

    auto perpDist = [](const std::array<double, 2>& p,
                       const std::array<double, 2>& a,
                       const std::array<double, 2>& b) -> double {
        double dx = b[0] - a[0], dy = b[1] - a[1];
        if (dx == 0 && dy == 0)
            return std::hypot(p[0] - a[0], p[1] - a[1]);
        double t = ((p[0] - a[0]) * dx + (p[1] - a[1]) * dy) / (dx * dx + dy * dy);
        t = std::clamp(t, 0.0, 1.0);
        return std::hypot(p[0] - (a[0] + t * dx), p[1] - (a[1] + t * dy));
    };

    // Iterative Douglas-Peucker using stack
    std::vector<bool> keep(coords.size(), false);
    keep[0] = true;
    keep[coords.size() - 1] = true;

    struct Range { size_t first, last; };
    std::vector<Range> stack;
    stack.push_back({0, coords.size() - 1});

    while (!stack.empty()) {
        auto [first, last] = stack.back();
        stack.pop_back();

        double maxD = 0;
        size_t maxI = first;
        for (size_t i = first + 1; i < last; ++i) {
            double d = perpDist(coords[i], coords[first], coords[last]);
            if (d > maxD) { maxD = d; maxI = i; }
        }
        if (maxD > tolerance) {
            keep[maxI] = true;
            if (maxI - first > 1) stack.push_back({first, maxI});
            if (last - maxI > 1) stack.push_back({maxI, last});
        }
    }

    std::vector<std::array<double, 2>> result;
    for (size_t i = 0; i < coords.size(); ++i) {
        if (keep[i])
            result.push_back({std::round(coords[i][0] * 1e5) / 1e5,
                              std::round(coords[i][1] * 1e5) / 1e5});
    }
    return result;
}

QString SettlementLoader::tileKey(const TileBounds& t)
{
    return QStringLiteral("%1,%2")
        .arg(std::get<0>(t), 0, 'f', 4)
        .arg(std::get<1>(t), 0, 'f', 4);
}

std::vector<TileBounds> SettlementLoader::gridTiles(double south, double west,
                                                     double north, double east)
{
    std::vector<TileBounds> tiles;
    double g = TILE_GRID;
    double s = std::floor(south / g) * g;
    while (s < north) {
        double wCur = std::floor(west / g) * g;
        while (wCur < east) {
            tiles.emplace_back(
                std::round(s * 1e4) / 1e4,
                std::round(wCur * 1e4) / 1e4,
                std::round((s + g) * 1e4) / 1e4,
                std::round((wCur + g) * 1e4) / 1e4
            );
            wCur += g;
        }
        s += g;
    }
    return tiles;
}

} // namespace vtol
